File: miscellaneous/graph-dfs/20230726/graph-dfs-rec.php
<?php
declare(strict_types=1);

function dfs(array $adjList, int $current, ?array &$seen=null, ?array &$values=null): array
{
    $seen = $seen ?: [$current];
    $values = $values ?: [$current];

    foreach ($adjList[$current] as $connection) {
        if (!in_array($connection, $seen)) {
            $seen[] = $connection;
            $values[] = $connection;
            dfs($adjList, $connection, $seen, $values);
        }
    }

    return $values;
}

File: miscellaneous/graph-dfs/20230726/graph_dfs_rec.py
from typing import List, Set


def dfs(adj_list: List[int], current: int, seen: Set[int] = None, values: List[int] = None) -> List[int]:
    values = values or [current]
    seen = seen or {0}

    for connection in adj_list[current]:
        if connection not in seen:
            seen.add(connection)
            values.append(connection)
            dfs(adj_list, connection, seen, values)

    return values